File: frontend/src/types/Invoice.ts
import { DiscountType } from '../constants/Enumeration/CommonEnum/DiscountEnum';
import { InvoiceStatus } from '../constants/Enumeration/InvoiceEnum/InvoiceEnum';
import { TCustomer } from './Customer';
import { TJob } from './Job';
import { TAddLineItem, TLineItem, TUpdateLineItem } from './LineItem';
import { TProperty } from './Property';

export enum PaymentMethod {
  Cash,
  Check,
  CardOnSite,
  Card,
  BankTransfer,
  Other,
}

export enum PaymentRecordStatus {
  Pending,
  Succeeded,
  Failed,
  Voided,
  Refunded,
}

export type TPayment = {
  id: string;
  invoiceId?: string | null;
  jobId?: string | null;
  amount: number;
  method: PaymentMethod;
  status: PaymentRecordStatus;
  processorReference?: string | null;
  paidAt?: string | null;
  recordedByUserId?: string | null;
  note?: string | null;
  createdAt: string;
  updatedAt: string;
};

export type TRecordInvoicePayment = {
  amount: number;
  method: PaymentMethod;
  paidAt: string;
  note?: string;
};

export type TInvoice = {
  id: string;
  customerId: string;
  customer: TCustomer;
  workspaceId: string;
  invoiceNumber: string;
  propertyId: string;
  property?: TProperty;
  jobId?: string;
  job?: TJob;
  title: string;
  lineItems: TLineItem[];
  taxRate: number;
  discount: number;
  discountType: DiscountType;
  subtotal: number;
  total: number;
  amountPaid: number;
  balanceDue: number;
  status: InvoiceStatus;
  issueDate: string;
  dueDate: string;
  paymentTerms: string;
  notes: string;
  internalNotes: string;
  isPaid: boolean;
  payments: TPayment[];
  createdAt: string;
  updatedAt: string;
};

export type TAddInvoice = {
  customerId: string;
  workspaceId: string;
  propertyId: string;
  jobId?: string;
  title: string;
  lineItems: TAddLineItem[];
  taxRate: number;
  discount: number;
  discountType: DiscountType;
  issueDate: string;
  dueDate?: string;
  paymentTerms: string;
  notes: string;
  internalNotes: string;
};

export type TUpdateInvoice = {
  id: string;
  title: string;
  lineItems: TUpdateLineItem[];
  taxRate?: number;
  discount?: number;
  discountType?: DiscountType;
  status?: InvoiceStatus;
  issueDate?: string;
  dueDate?: string;
  paymentTerms?: string;
  notes?: string;
  internalNotes?: string;
};

export type TInvoiceFilter = {
  q?: string;
  sortBy?: string;
  sort?: string;
  status?: string;
  dueDateMin?: string;
  dueDateMax?: string;
  totalMin?: number;
  totalMax?: number;
};

export type TInvoiceStats = {
  totalOutstanding: number;
  totalPaidThisMonth: number;
  overdueCount: number;
  averageInvoiceValue: number;
};

File: backend/Models/Invoice.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Text.Json.Serialization;
using backend.Models.QuoteModels;
using backend.Services.Billing;

namespace backend.Models;

public class Invoice
{
    [Key]
    public Guid Id { get; set; }
    public Guid WorkspaceId { get; set; }
    public Guid CustomerId { get; set; }
    public Customer? Customer { get; set; }
    public string InvoiceNumber { get; set; } = string.Empty;

    public Guid PropertyId { get; set; }
    public Property? Property { get; set; }

    public Guid? JobId { get; set; }
    public Job? Job { get; set; }

    public string Title { get; set; } = string.Empty;
    public List<LineItem> LineItems { get; set; } = [];
    public List<Payment> Payments { get; set; } = [];
    public decimal TaxRate { get; set; }
    public decimal Discount { get; set; }
    public DiscountType DiscountType { get; set; } = DiscountType.FixedAmount;
    private decimal _subtotal;
    private decimal _discountAmount;
    private decimal _taxAmount;
    private decimal _total;
    public decimal Subtotal { get => EffectiveTotals().Subtotal; private set => _subtotal = value; }
    public decimal DiscountAmount { get => EffectiveTotals().Discount; private set => _discountAmount = value; }
    public decimal TaxAmount { get => EffectiveTotals().TaxAmount; private set => _taxAmount = value; }
    public decimal Total { get => EffectiveTotals().Total; private set => _total = value; }
    [NotMapped]
    public decimal AmountPaid => PaymentLedgerCalculator.CalculateAmountPaid(Payments);
    [NotMapped]
    public decimal BalanceDue => PaymentLedgerCalculator.CalculateBalanceDue(Total, Payments);
    [NotMapped]
    public bool IsPaid => BalanceDue <= 0m;
    [JsonIgnore]
    [Column("Status")]
    public InvoiceStatus WorkflowStatus { get; set; } = InvoiceStatus.Draft;
    [NotMapped]
    public InvoiceStatus Status => PaymentLedgerCalculator.DeriveInvoiceStatus(WorkflowStatus, BalanceDue, AmountPaid, DueDate, DateTime.UtcNow);
    public DateTime? SentAt { get; set; }
    public DateTime IssueDate { get; set; } = DateTime.UtcNow;
    public DateTime DueDate { get; set; }
    public string PaymentTerms { get; set; } = string.Empty;
    public string Notes { get; set; } = string.Empty;
    public string InternalNotes { get; set; } = string.Empty;
    public bool IsArchived { get; set; }
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
    [Timestamp]
    public byte[] RowVersion { get; set; } = [];

    public void RecalculateTotals()
    {
        var totals = TotalsCalculator.Calculate(LineItems, DiscountType, Discount, TaxRate);
        _subtotal = totals.Subtotal;
        _discountAmount = totals.Discount;
        _taxAmount = totals.TaxAmount;
        _total = totals.Total;
    }

    private TotalsBreakdown EffectiveTotals()
        => LineItems.Count > 0
            ? TotalsCalculator.Calculate(LineItems, DiscountType, Discount, TaxRate)
            : new TotalsBreakdown(_subtotal, _discountAmount, 0m, _taxAmount, _total);
}

public enum InvoiceStatus
{
    Draft = 0,
    Sent = 1,
    Paid = 2,
    Overdue = 3,
    Partial = 4
}
